scipy/ndimage/src/_ni_label.pyx

```diff
@@ -326,9 +326,9 @@ cpdef _label(np.ndarray input,
                 # Take neighbor labels
                 PyArray_ITER_RESET(itstruct)
                 for ni in range(num_neighbors):
-                    neighbor_use_prev = (<np.int_t *> PyArray_ITER_DATA(itstruct))[0]
-                    neighbor_use_adjacent = (<np.int_t *> (<char *> PyArray_ITER_DATA(itstruct) + ss))[0]
-                    neighbor_use_next = (<np.int_t *> (<char *> PyArray_ITER_DATA(itstruct) + 2 * ss))[0]
+                    neighbor_use_prev = (<np.npy_long *> PyArray_ITER_DATA(itstruct))[0]
+                    neighbor_use_adjacent = (<np.npy_long *> (<char *> PyArray_ITER_DATA(itstruct) + ss))[0]
+                    neighbor_use_next = (<np.npy_long *> (<char *> PyArray_ITER_DATA(itstruct) + 2 * ss))[0]
```

Comment:
This isn't quite correct unfortunately, because above `dtype=int` is used, which makes this wrong.

It seems to me that the input here is actually always a boolean, so maybe its best to just cast to bool at the top (which it should already be) and use `npy_bool` (or similar) here.

Reply:
I pushed in a commit that seems to pass full test suite with NumPy `1.23.x` and latest NumPy `main`.

That said, I'm really confused, when I use `np.long` as `dtype` on NumPy `1.23.x` I see:

> DeprecationWarning: `np.long` is a deprecated alias for `np.compat.long`. To silence this warning, use `np.compat.long` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `int` 

That warning to me with NumPy `1.23.x` suggests that I can use `dtype=int` to safely "type-transmit" what `np.long` was doing before? That's why I went for `npy_long` here, on the assumption that built-in `int` for dtype-specifier is a safe replacement for `np.long` which == `npy_long`. Otherwise, what is going on with that warning? That's the second bullet point in my original comment above too, so clearly I've misunderstood something... I believe I changed to `npy_bool` now, but I'm not fully convinced things are making sense to me.

Reply:
Confusing, I guess... `np.long` used to be an alias to `int` (after/due to Python 2->3) and `int` would previously go to C-long, so `np.long == npy_long` was only true when long was the default integer.

So `np.long` would actually work, but the problem is that we didn't (and still don't) want it to be an alias for Python `int`.  That leaves only `"long"` the string to really access the long type.  Maybe we need to make it clearer that the only good way is unfortunately via the string.